h2o-core/src/main/java/hex/grid/GridSearch.java

```diff
@@ -205,26 +208,40 @@ public void onBuildFailure(final ParallelModelBuilder.ModelBuildFailure modelBui
       } finally {
         parallelSearchGridLock.unlock();
       }
-      attemptBuildNextModel(parallelModelBuilder, null);
+    }
+
+    @Override
+    public void afterBuildProcessing(ParallelModelBuilder parallelModelBuilder, Model finishedModel) {
+      attemptBuildNextModel(parallelModelBuilder, finishedModel);
     }
 
     private void attemptBuildNextModel(final ParallelModelBuilder parallelModelBuilder, final Model previousModel) {
-      // Attempt to train next model
-      final MP nextModelParams = getNextModelParams(hyperspaceIterator, previousModel, grid);
-      if (nextModelParams != null
-              && isThereEnoughTime()
-              && !_job.stop_requested()
-              && !_hyperSpaceWalker.stopEarly(previousModel, grid.getScoringInfos())) {
 
-        // FIXME: ModelFeeder's <MP> type parameter is hiding GridSearch's one. Ideally would have moved ModelFeeder outside but it is too coupled with GridSearch class.
-        reconcileMaxRuntime(grid._key, hyperspaceIterator, nextModelParams);
+      synchronized (hyperspaceIterator) {
+        if (hyperspaceIterator.hasNext(previousModel)
+                && isThereEnoughTime()
+                && !hasReachedMaxModels(parallelModelBuilder)
+                && !_job.stop_requested()
+                && !_hyperSpaceWalker.stopEarly(previousModel, grid.getScoringInfos())) {
 
-        parallelModelBuilder.run(Collections.singletonList(ModelBuilder.make(nextModelParams)));
-      } else {
-        parallelModelBuilder.noMoreModels();
+          final MP nextModelParams = getNextModelParams(hyperspaceIterator, previousModel, grid);
+
+          // FIXME: ModelFeeder's <MP> type parameter is hiding GridSearch's one. Ideally would have moved ModelFeeder outside but it is too coupled with GridSearch class.
+          reconcileMaxRuntime(grid._key, hyperspaceIterator, nextModelParams);
+
+          parallelModelBuilder.run(Collections.singletonList(ModelBuilder.make(nextModelParams)));
+        } else {
+          parallelModelBuilder.noMoreModels();
+        }
       }
     }
 
+    private boolean hasReachedMaxModels(ParallelModelBuilder parallelModelBuilder) {
```

Comment:
just an idea: wouldnt this be better as a method on parallelModelBuilder with max_models parameter?

Reply:
@honzasterba with existing design `ParallelModelBuilder' does not have reference to `hyperspaceIterator`...it is working with `Collection<ModelBuilder>`.

Reply:
@honzasterba sorry, with max_model parameter it could work. Will try